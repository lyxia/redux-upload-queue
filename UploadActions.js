//@flow

import {createAction} from 'redux-actions'

import UploadEventEmitter from './UploadEventEmitter'

import {config} from './UploadComponent'

export const pushUploadItem = createAction('Upload/pushItem')
export const deleteUploadItem = createAction('Upload/deleteItem')
export const cleanUpload = createAction('Upload/cleanUpload')
export const startuploadItem = createAction('Upload/startUploadItem')
export const uploadItemFailed = createAction('Upload/failedUploadItem')
export const uploadItemSuccess = createAction('Upload/successUploadItem')

export const startUpload = createAction('Upload/startUpload')
export const uploadComplete = createAction('Upload/uploadComplete')

export const uploadCancle = createAction('Upload/uploadCancle')
export const registerUpload = createAction('Upload/registerUpload')
export const destroyUpload = createAction('Upload/destroyUpload')

export function upload(name, config) {
    return (dispatch, getState) => {
        const state = getState()
        if(state.upload.get(name)) {
            if(!state.upload.get(name).uploading) {
                //开始上传
                dispatch(startUpload({upload: name}))
                return uploadNextItem(name, dispatch, getState, config)
            }
        }
        return Promise.resolve()
    }
}

function uploadItem(name, item, dispatch, getState, config) {
    dispatch(startuploadItem({upload: name}))
    return config.uploadApi(item.filepath, item.name)
        .then((model) => {
            if(config.isSuccess(model)) {
                //fileName
                dispatch(uploadItemSuccess({upload: name, id:config.getUUIDFuc(model)}))
            } else {
                dispatch(uploadItemFailed({upload: name}))
            }
            return uploadNextItem(name, dispatch, getState, config)
        })
        .catch((error) => {
            dispatch(uploadItemFailed({upload: name}))
            return uploadNextItem(name, dispatch, getState, config)
        })
}

function getNextUploadItem(name, getState) {
    const state = getState()
    const upload = state.upload.get(name)
    if(!upload || upload.cancle) return null

    return upload.wattingQueue.first()
}

function uploadNextItem(name, dispatch, getState, config) {
    let nextItem = getNextUploadItem(name, getState)
    if(nextItem) {
        return uploadItem(name, nextItem, dispatch, getState, config)
    } else {
        dispatch(uploadComplete({upload: name}))
        //发送事件
        UploadEventEmitter.emitComplete(name)
        return Promise.resolve()
    }
}