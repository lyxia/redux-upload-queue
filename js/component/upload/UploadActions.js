//@flow

import {createAction} from 'redux-actions'

import {uploadImageWithPost} from '../../networking/FetchData'

import UploadEventEmitter from './UploadEventEmitter'

export const pushUploadItem = createAction('Upload/pushItem')
export const deleteUploadItem = createAction('Upload/deleteItem')
export const startuploadItem = createAction('Upload/startUploadItem')
export const uploadItemFailed = createAction('Upload/failedUploadItem')
export const uploadItemSuccess = createAction('Upload/successUploadItem')

export const startUpload = createAction('Upload/startUpload')
export const uploadComplete = createAction('Upload/uploadComplete')

export const uploadCancle = createAction('Upload/uploadCancle')
export const registerUpload = createAction('Upload/registerUpload')
export const destroyUpload = createAction('Upload/destroyUpload')

export function upload(name) {
    return (dispatch, getState) => {
        const state = getState()
        if(state.upload.get(name)) {
            if(!state.upload.get(name).uploading) {
                //开始上传
                dispatch(startUpload({upload: name}))
                uploadNextItem(name, dispatch, getState)
            }
        }
    }
}

function uploadItem(name, item, dispatch, getState) {
    dispatch(startuploadItem({upload: name}))
    uploadImageWithPost(item.filepath, item.name)
        .then((model) => {
            if(model.error == -1) {
                dispatch(uploadItemSuccess({upload: name, id:model.fileName}))
            } else {
                dispatch(uploadItemFailed({upload: name}))
            }
            uploadNextItem(name, dispatch, getState)
        })
        .catch((error) => {
            dispatch(uploadItemFailed({upload: name}))
            uploadNextItem(name, dispatch, getState)
        })
}

function getNextUploadItem(name, getState) {
    const state = getState()
    const upload = state.upload.get(name)
    if(!upload || upload.cancle) return null

    return upload.wattingQueue.first()
}

function uploadNextItem(name, dispatch, getState) {
    let nextItem = getNextUploadItem(name, getState)
    if(nextItem) {
        uploadItem(name, nextItem, dispatch, getState)
    } else {
        dispatch(uploadComplete({upload: name}))
        //发送事件
        UploadEventEmitter.emitComplete(name)
    }
}