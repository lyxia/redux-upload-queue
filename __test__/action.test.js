import * as UploadActions from '../UploadActions'
import config, {store} from './UploadConfig'

afterEach(() => {
  store.clearActions()
  fetch.resetMocks()
})

it('upload one file success action test', () => {
  fetch.mockResponseOnce(JSON.stringify({ error: null, id: '123456' }))

  store.dispatch(UploadActions.registerUpload({upload: 'uploadKey'}))
  store.dispatch(UploadActions.pushUploadItem({upload: 'uploadKey', name: 'fileOne', filePath: 'filePathOne'}))
  return store.dispatch(UploadActions.upload('uploadKey', config))
          .then(() => {
            expect(store.getActions()).toMatchSnapshot()
          })
})

it('upload one file fail action test', () => {
  fetch.mockResponseOnce(JSON.stringify({ error: new Error('fail') }))

  store.dispatch(UploadActions.registerUpload({upload: 'uploadKey'}))
  store.dispatch(UploadActions.pushUploadItem({upload: 'uploadKey', name: 'fileOne', filePath: 'filePathOne'}))
  return store.dispatch(UploadActions.upload('uploadKey', config))
          .then(() => {
            expect(store.getActions()).toMatchSnapshot()
          })
})

it('upload mult success action test', () => {
  fetch.mockResponse(JSON.stringify({ error: null, id: '123456' }))

  store.dispatch(UploadActions.registerUpload({upload: 'uploadKey'}))
  store.dispatch(UploadActions.pushUploadItem({upload: 'uploadKey', name: 'fileOne', filePath: 'filePathOne'}))
  store.dispatch(UploadActions.pushUploadItem({upload: 'uploadKey', name: 'fileTwo', filePath: 'filePathTwo'}))
  store.dispatch(UploadActions.pushUploadItem({upload: 'uploadKey', name: 'fileThree', filePath: 'filePathThree'}))
  return store.dispatch(UploadActions.upload('uploadKey', config))
          .then(() => {
            expect(store.getActions()).toMatchSnapshot()
          })
})

it('upload mult fail and reupload action test', () => {
  fetch.mockResponses(
    [
      JSON.stringify({ error: null, id: '123456' })
    ],
    [
      JSON.stringify({ error: null, id: '123456' })
    ],
    [
      JSON.stringify({ error: new Error('fail') })
    ],
    [
      JSON.stringify({ error: null, id: '123456' })
    ],
  )

  store.dispatch(UploadActions.registerUpload({upload: 'uploadKey'}))
  store.dispatch(UploadActions.pushUploadItem({upload: 'uploadKey', name: 'fileOne', filePath: 'filePathOne'}))
  store.dispatch(UploadActions.pushUploadItem({upload: 'uploadKey', name: 'fileTwo', filePath: 'filePathTwo'}))
  store.dispatch(UploadActions.pushUploadItem({upload: 'uploadKey', name: 'fileThree', filePath: 'filePathThree'}))
  return store.dispatch(UploadActions.upload('uploadKey', config))
          .then(() => {
            return store.dispatch(UploadActions.upload('uploadKey', config))
          })
          .then(() => {
            expect(store.getActions()).toMatchSnapshot()
          })
})
