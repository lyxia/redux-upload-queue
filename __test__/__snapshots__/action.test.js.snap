// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload mult fail and reupload action test 1`] = `
Array [
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/registerUpload",
  },
  Object {
    "payload": Object {
      "filePath": "filePathOne",
      "name": "fileOne",
      "upload": "uploadKey",
    },
    "type": "Upload/pushItem",
  },
  Object {
    "payload": Object {
      "filePath": "filePathTwo",
      "name": "fileTwo",
      "upload": "uploadKey",
    },
    "type": "Upload/pushItem",
  },
  Object {
    "payload": Object {
      "filePath": "filePathThree",
      "name": "fileThree",
      "upload": "uploadKey",
    },
    "type": "Upload/pushItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUpload",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUploadItem",
  },
  Object {
    "payload": Object {
      "id": "123456",
      "upload": "uploadKey",
    },
    "type": "Upload/successUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUploadItem",
  },
  Object {
    "payload": Object {
      "id": "123456",
      "upload": "uploadKey",
    },
    "type": "Upload/successUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/failedUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/uploadComplete",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUpload",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUploadItem",
  },
  Object {
    "payload": Object {
      "id": "123456",
      "upload": "uploadKey",
    },
    "type": "Upload/successUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/uploadComplete",
  },
]
`;

exports[`upload mult success action test 1`] = `
Array [
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/registerUpload",
  },
  Object {
    "payload": Object {
      "filePath": "filePathOne",
      "name": "fileOne",
      "upload": "uploadKey",
    },
    "type": "Upload/pushItem",
  },
  Object {
    "payload": Object {
      "filePath": "filePathTwo",
      "name": "fileTwo",
      "upload": "uploadKey",
    },
    "type": "Upload/pushItem",
  },
  Object {
    "payload": Object {
      "filePath": "filePathThree",
      "name": "fileThree",
      "upload": "uploadKey",
    },
    "type": "Upload/pushItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUpload",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUploadItem",
  },
  Object {
    "payload": Object {
      "id": "123456",
      "upload": "uploadKey",
    },
    "type": "Upload/successUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUploadItem",
  },
  Object {
    "payload": Object {
      "id": "123456",
      "upload": "uploadKey",
    },
    "type": "Upload/successUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUploadItem",
  },
  Object {
    "payload": Object {
      "id": "123456",
      "upload": "uploadKey",
    },
    "type": "Upload/successUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/uploadComplete",
  },
]
`;

exports[`upload one file fail action test 1`] = `
Array [
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/registerUpload",
  },
  Object {
    "payload": Object {
      "filePath": "filePathOne",
      "name": "fileOne",
      "upload": "uploadKey",
    },
    "type": "Upload/pushItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUpload",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/failedUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/uploadComplete",
  },
]
`;

exports[`upload one file success action test 1`] = `
Array [
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/registerUpload",
  },
  Object {
    "payload": Object {
      "filePath": "filePathOne",
      "name": "fileOne",
      "upload": "uploadKey",
    },
    "type": "Upload/pushItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUpload",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/startUploadItem",
  },
  Object {
    "payload": Object {
      "id": "123456",
      "upload": "uploadKey",
    },
    "type": "Upload/successUploadItem",
  },
  Object {
    "payload": Object {
      "upload": "uploadKey",
    },
    "type": "Upload/uploadComplete",
  },
]
`;
